client/node-common/src/service.rs

```diff
@@ -0,0 +1,750 @@
+            sc_service::new_full_parts::<BlockOf<T>, RuntimeApiOf<T>, _>(
+                parachain_config,
+                telemetry.as_ref().map(|(_, telemetry)| telemetry.handle()),
+                executor,
+            )?;
+        let client = Arc::new(client);
+
+        let telemetry_worker_handle = telemetry.as_ref().map(|(worker, _)| worker.handle());
+
+        let telemetry = telemetry.map(|(worker, telemetry)| {
+            task_manager
+                .spawn_handle()
+                .spawn("telemetry", None, worker.run());
+            telemetry
+        });
+
+        let transaction_pool = sc_transaction_pool::BasicPool::new_full(
+            parachain_config.transaction_pool.clone(),
+            parachain_config.role.is_authority().into(),
+            parachain_config.prometheus_registry(),
+            task_manager.spawn_essential_handle(),
+            client.clone(),
+        );
+
+        Ok(Self {
+            client,
+            backend,
+            transaction_pool,
+            telemetry,
+            telemetry_worker_handle,
+            task_manager,
+            keystore_container,
+            hwbench,
+            prometheus_registry: parachain_config.prometheus_registry().cloned(),
+            network: TypeIdentity::from_type(()),
+            tx_handler_controller: TypeIdentity::from_type(()),
+            import_queue_service: TypeIdentity::from_type(()),
+        })
+    }
+}
+
+impl<T: Config, SNetwork, STxHandler, SImportQueueService>
+    NodeBuilder<T, SNetwork, STxHandler, SImportQueueService>
+where
+    BlockOf<T>: cumulus_primitives_core::BlockT,
+    ParachainNativeExecutorOf<T>: NativeExecutionDispatch + 'static,
+    RuntimeApiOf<T>: ConstructRuntimeApi<BlockOf<T>, ClientOf<T>> + Sync + Send + 'static,
+    ConstructedRuntimeApiOf<T>: TaggedTransactionQueue<BlockOf<T>>
+        + BlockBuilder<BlockOf<T>>
+        + cumulus_primitives_core::CollectCollationInfo<BlockOf<T>>,
+{
+    #[must_use]
+    pub async fn build_relay_chain_interface(
+        &mut self,
+        parachain_config: &Configuration,
+        polkadot_config: Configuration,
+        collator_options: CollatorOptions,
+    ) -> sc_service::error::Result<(
+        Arc<(dyn RelayChainInterface + 'static)>,
+        Option<CollatorPair>,
+    )> {
+        build_relay_chain_interface(
+            polkadot_config,
+            &parachain_config,
+            self.telemetry_worker_handle.clone(),
+            &mut self.task_manager,
+            collator_options.clone(),
+            self.hwbench.clone(),
+        )
+        .await
+        .map_err(|e| sc_service::Error::Application(Box::new(e) as Box<_>))
+    }
+
+    /// Given an import queue, calls `cumulus_client_service::build_network` and
+    /// stores the returned objects in `self.network` and `self.tx_handler_controller`.
+    ///
+    /// Can only be called once on a `NodeBuilder` that doesn't have yet network
+    /// data.
```

Comment:
I really like this, I knew it can be done using const generics and a state machine enum, but I didn't know about the `TypeIdentity` trick, will need to look into it.

Reply:
This is the "Typestate Pattern" described here: http://cliffle.com/blog/rust-typestate/

Usually you need an `impl` block per typestate, but `core_extensions::TypeIdentity` allows to state that a type parameter is equal to a concrete type and convert between both, which allow to have a single `impl` block with `where` clauses on the functions. (except for `new`, as it doesn't take `self` Rust is unable to infer which type to use, even if there is only 1 type `T` implementing `TypeIdentity<Type = T>`)